python_modules/libraries/dagster-k8s/dagster_k8s/pipes.py

```diff
@@ -421,3 +651,93 @@ def build_pod_body(
             "spec": spec,
         },
     )
+
+
+def _process_log_stream(stream: Iterator[bytes]) -> Iterator[LogItem]:
+    """This expects the logs to be of the format b'<timestamp> <msg>' and only the
+    '<msg>' is forwarded to Dagster. If the <timestamp> is not there then the lines
+    will be joined together. There is a limitation that the first item in the stream
+    needs to always contain a timestamp as the first element.
+
+    The timestamp is expected to be in '2024-03-22T02:17:29.885548Z' format and
+    if the subsecond part will be truncated to microseconds.
+
+    If we fail parsing the timestamp, then the priority will be set to zero in
+    order to not drop any log items.
+
+    Args:
+        stream (Iterator[bytes]): A stream of log chunks
+
+    Yields:
+        Iterator[LogItem]: A log containing the timestamp and msg
+    """
+    timestamp = ""
+    log = ""
+
+    for log_chunk in stream:
+        for line in log_chunk.decode("utf-8").split("\n"):
+            maybe_timestamp, _, tail = line.partition(" ")
+            if not timestamp:
+                # The first item in the stream will always have a timestamp.
+                timestamp = maybe_timestamp
+                log = tail
+            elif maybe_timestamp == timestamp:
+                # We have multiple messages with the same timestamp in this chunk, add them separated
+                # with a new line
+                log += f"\n{tail}"
+            elif not (
+                len(maybe_timestamp) == len(timestamp) and _is_kube_timestamp(maybe_timestamp)
+            ):
+                # The line is continuation of a long line that got truncated and thus doesn't
+                # have a timestamp in the beginning of the line.
+                # Since all timestamps in the RFC format returned by Kubernetes have the same
+                # length (when represented as strings) we know that the value won't be a timestamp
+                # if the string lengths differ, however if they do not differ, we need to parse the
+                # timestamp.
+                log += line
+            else:
+                # New log line has been observed, send in the next cycle
+                yield LogItem(timestamp=timestamp, log=log)
+                timestamp = maybe_timestamp
+                log = tail
+
+    # Send the last message that we were building
+    if log or timestamp:
+        yield LogItem(timestamp=timestamp, log=log)
+
+
+def _is_kube_timestamp(maybe_timestamp: str) -> bool:
+    # fromisoformat only works properly in Python 3.11+
+    # Once pre 3.11 backwards compatibility is dropped
+    # anything after this if statement can be deleted
+    if sys.version_info >= (3, 11):
+        try:
+            datetime.fromisoformat(maybe_timestamp)
+            return True
+        except ValueError:
+            return False
```

Comment:
Not checking the `sys.version_info` every time we have a log could make a difference. Just defining a different function based on the Python version is more elegant, I think:
```suggestion
if sys.version_info >= (3, 11):
    # fromisoformat only works properly in Python 3.11+
    def _is_kube_timestamp(maybe_timestamp: str) -> bool:
        try:
            datetime.fromisoformat(maybe_timestamp)
            return True
        except ValueError:
            return False
else:
    # Once pre 3.11 backwards compatibility is dropped
    # anything after this if statement can be deleted
    def _is_kube_timestamp(maybe_timestamp: str) -> bool:
```

Reply:
It could be also nice to add a few tests to ensure that the function works.

Reply:
Added some tests in `test_pipe_log_reader.py`. It's a little strange because the tests could change based on environment (Python 3.11). Though the specific test cases should be agnostic to each env, and also Dagster target env is 3.10 right now.

Note that we only ensure correctness for RFC3339Nano timestamp, since that's what Docker runtime uses. Not sure if this would work with containerd for instance.